### ISSUE TITLE
Make copr-builds from the merged result not the contributor's branch

### ISSUE THREAD

**Issue by yarda** [packit/packit-service#245]
I am not sure whether this is the correct tracker (sorry if it isn't :). This problem was originally pointed to me by @olysonek. According to the git hashes from the log it seems Packit runs the copr-build from the contributor's branch, but as most CI do, I think it should run the copr-build from the merged result - i.e. try to merge the contributor's branch to the master and make the copr-build from the result. It should check that the merge will not break the project, not the contributor's branch.

**Comment by TomasTomecek** [comment]
> I am not sure whether this is the correct tracker (sorry if it isn't :).

Don't worry about it, we can easily move b/w our trackers. We are glad you are giving us all the feedback.


> This problem was originally pointed to me by @olysonek. According to the git hashes from the log it seems Packit runs the copr-build from the contributor's branch

Yes it this, we are not doing any git magic.

> but as most CI do, I think it should run the copr-build from the merged result - i.e. try to merge the contributor's branch to the master and make the copr-build from the result. It should check that the merge will not break the project, not the contributor's branch.

I'd say this is a good suggestion and probably worth of making it the default behaviour.

On the other hand, I can imagine a scenario when I, as a contributor, am working on a change and packit "breaks" it by performing the merge while I want to see only test results for the tests affected by my change. But, on the other hand, the rebase is inevitable so... no excuses.

What do others think?

**Comment by olysonek** [comment]
> On the other hand, I can imagine a scenario when I, as a contributor, am working on a change and packit "breaks" it by performing the merge while I want to see only test results for the tests affected by my change. But, on the other hand, the rebase is inevitable so... no excuses.

I'd like to point out that a rebase is not inevitable - the project maintainer can always perform a manual merge, fixing up any incompatibilities between the two branches in the merge commit. That is in fact one of the responsibilities of a project maintainer (at least as regarded in some social circles).

I think it would make sense for Packit to support testing both the contributor's branch and the merge. The results from the contributor's branch would be primarily targeted at the contributor, informing them whether the changes they made are sane. The results from the merge would be primarily targeted at the project maintainer, informing them whether merging the changes will break the project or not.

However, testing both by default in all projects would most likely be a waste of resources. The default should be just one of these options - the one that makes most sense in the context of a pull request - testing the merge, informing the maintainer whether it's safe to pull the changes or not.

**Comment by lachmanfrantisek** [comment]
We still need to discuss that.

**Comment by jpopelka** [comment]
It should be configurable with a default being "not-merging".

**Comment by olysonek** [comment]
Bump.

**Comment by olysonek** [comment]
Bump.

**Comment by lachmanfrantisek** [comment]
@yarda @olysonek We're finishing the support in our API library and are thinking about the support in the service. Here are some points from the implementation perspective:

* Doing both is hard for us to implement and probably a waste of resources as you've pointed out.
* Making it configurable is not so easy -- we need to know the git reference for loading packit config. I am worried that having a different ref of config and code can cause only confusion and problems. And reloading of config would take a non-trivial time to do.

So, it looks like the only easily-doable option is to turn this on for all. I know you are requesting this option, but do you think it will be a problem for other teams?

And here are the tasks that need to be done here:
* [ ] correctly receive events (merge checks take some time)
* [ ] give the user a warning when merge check fail (ask for rebase and don't do anything)
* [ ] load the config from the correct ref and checkout the correct ref of the code

**Comment by yarda** [comment]
@lachmanfrantisek thanks for info. I can't speak for other teams and I even don't know how they use packit, but for me as a maintainer it is important to know whether the PR will or will not break the project. So for me the proposed behaviour is definitely beneficial and I think for other maintainers working the similar way it may be also beneficial.

Adding @CZerta who now works on the TuneD project, @olysonek is no more on the TuneD project, but he may be still interested in the feature.

**Comment by olysonek** [comment]
Sorry guys, no time for this right now.

But I guess the only projects where this could be a problem could be those where merges are frequently non-trivial. I.e. where it is a common occurrence that the maintainer has to resolve merge conflicts by hand or decide which branch(es) the change should be merged into. In such cases it might be more beneficial to test the contributor's branch and let the maintainer deal with the rest. But what do I know - just brainstorming. I can't think of any specific examples; maybe kernel. But I don't think they use Packit.

Good luck and take care. :)

**Comment by lachmanfrantisek** [comment]
Related discussion: https://github.com/packit/packit/discussions/1310

**Comment by martinpitt** [comment]
FTR, in cockpit we've been playing around with using GitHub's automatic merge commits (`github.event.pull_request.merge_commit_sha`) are a bit of a disaster. Quoting our [recent commit](https://github.com/cockpit-project/cockpit/commit/678f910515b1dee09b076267173288466382cecd):

> Using the .merge_commit_sha from GitHub is seriously dangerous (it's usually not up to date when the workflow gets launched) and kinda pointless (it's completely transient, and will change next time someone pushes main).

So in that above commit we reverted using the merge commit, went back to the original `github.event.pull_request.head.sha` , and calling `git merge` ourselves.

So maybe you want to do that in packit, after all.. Sorry for the back and forth, we only learned that the hard way! (Thanks to @
@allisonkarlitskaya for battling with this)
